### ISSUE TITLE
logictest: JSON payload for failures do not include the test name, causing TC to omit the failure in the overview

### ISSUE THREAD

**Issue by yuzefovich** [cockroachdb/cockroach#57516]
[Here](https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_CockroachMetamorphicNightly/2493436:id/logTestLogic547935272/logictesttest.log) is an example of the log from the logic test run that contains a crash (the panic was caught and logged, and then the process crashed). However, the [build](https://teamcity.cockroachdb.com/viewLog.html?buildId=2493436&buildTypeId=Cockroach_Nightlies_CockroachMetamorphicNightly&tab=buildResultsDiv&branch_Cockroach=57483) doesn't show anything about `pkg/sql/logictest` package that actually experienced the crash. This needs more investigation.

Jira issue: CRDB-3525

**Comment by knz** [comment]
So after looking I think this is the problem:

```
{"Time":"2020-12-03T14:10:58.994491054-05:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestLogic/local/alter_type_owner","Output":"\t/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/table_cache.go:167 +0x1a5\n
"}
{"Time":"2020-12-03T14:10:59.289223911-05:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Output":"FAIL\tgithub.com/cockroachdb/cockroach/pkg/sql/logictest\t21.017s\n"}
{"Time":"2020-12-03T14:10:59.28928293-05:00","Action":"fail","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Elapsed":21.017}
```

The lines with "Test" do not contain an `"Action":"fail"`. Only the last line for the entire package is marked with `fail`, but that line does not contain the `Test` field any more.

Here is another one of the failing tests in this run:

```
{"Time":"2020-12-03T14:05:42.578930961-05:00","Action":"fail","Package":"github.com/cockroachdb/cockroach/pkg/col/coldata","Test":"TestSwapNulls/TestSwapNullWithNotNull","Elapsed":0}
```

As you see the `"Action":"fail"` carries the name of the test.

So the root problem is that for logic tests, the "fail" lines are reported for the entire package and not for individual tests. This is a bug in the JSON output generator for logic tests.

----

More details

I can see two panics in the JSON output:

```
{"Time":"2020-12-03T14:10:57.712274214-05:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestLogic/local/and_or","Output":"E201203 19:10:57.702280 211632 util/log/logcrash/crash_reporting.go:140  [n1,client=‹127.0.0.1:39336›,hostssl,user=ro
ot] a panic has occurred!\n"}
{"Time":"2020-12-03T14:10:57.712334654-05:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestLogic/local/and_or","Output":"runtime error: index out of range [0] with length 0\n"}
{"Time":"2020-12-03T14:10:57.712346398-05:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestLogic/local/and_or","Output":"(1) attached stack trace\n"}
{"Time":"2020-12-03T14:10:57.712353433-05:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestLogic/local/and_or","Output":"  -- stack trace:\n"}
...
```

and

```
{"Time":"2020-12-03T14:10:57.728271074-05:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestLogic/local/alter_type_owner","Output":"--- progress: testdata/logic_test/alter_primary_key: 33 statements/queries\n"}
{"Time":"2020-12-03T14:10:57.74722522-05:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestLogic/local/alter_type_owner","Output":"panic: runtime error: index out of range [0] with length 0 [recovered]\n"}
{"Time":"2020-12-03T14:10:57.747283855-05:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestLogic/local/alter_type_owner","Output":"\tpanic: runtime error: index out of range [0] with length 0\n"}
{"Time":"2020-12-03T14:10:57.747293993-05:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestLogic/local/alter_type_owner","Output":"\n"}
{"Time":"2020-12-03T14:10:57.747300427-05:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestLogic/local/alter_type_owner","Output":"goroutine 211632 [running]:\n"}
...
```

Finishing all the logic tests with:

```
{"Time":"2020-12-03T14:10:59.289223911-05:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Output":"FAIL\tgithub.com/cockroachdb/cockroach/pkg/sql/logictest\t21.017s\n"}
```

The second panic goroutine dump is humongous btw - maybe a megabyte or so? 
Even if we solve the root problem above, I think TC has a built-in limit that will prevent showing extremely large error traces.

**Comment by knz** [comment]
The thing is I am not an expert with the logic test runner (any more), neither do I understand the JSON output generator nor the output parser.

Who are the best people who understand these things nowadays? @tbg do you know?

**Comment by knz** [comment]
I am unassigning myself - this issue is not related to logging nor the recent changes around `log.Scope`. It's possible we're seeing an after-effect of the upgrade to TC, or Go 1.15 (the test output in 1.13 was generated differently).

**Comment by knz** [comment]
I'd also suggest someone from SQL experience to look at this. Perhaps it's a question of placing a `t.Error()` call in the right place in `logic.go`.

**Comment by tbg** [comment]
The JSON output is standard `go test -json` and we are using TeamCity's built in parser. From the stack trace, it looks like a logic test was crashing when tests were running concurrently. In that case, it's usually impossible to blame a particular test for the failure. The "Go standard" (as reflected by what `go test -json` does with this) is that this is tracked as a package-level failure, and not pinned on any individual test. This makes sense, when you think about it - it's hard to do much better; the only other option is to mark all inflight tests as failed which is similarly not useful in cases like logic tests, where this number can be large.

To mitigate that, we have the "un-jsoned" output in the artifacts:

![image](https://user-images.githubusercontent.com/5076964/101153222-6d92ec80-3624-11eb-8671-58d9714480d3.png)

The bottom line is that Go just doesn't have enough isolation between parallel tests to get us the reporting that we want. We need to avoid unhandled panics on all goroutines by turning them into `t.Error` and winding down the test gracefully. This is, unfortunately, annoying.

**Comment by knz** [comment]
I think we would be happy even if we just saw the top level name `TestLogic` in the TC output, even if we can't distinguish the specific test that failed. 

Heck, we could be happy if TC just bothered to print out the name of the Go package `logictest`, even without a test name.

After all, TC *does* see this line:

```
{"Time":"2020-12-03T14:10:59.289223911-05:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Output":"FAIL\tgithub.com/cockroachdb/cockroach/pkg/sql/logictest\t21.017s\n"}
```

I think what I am not happy about is that the Go test runner sees the package failing (FAIL here) but does not include `"Action":"fail"`!! This looks like a bug to me.

I could go and muck with the logging package, to translate calls to `log.Fatal()` under a `TestLogScope` into a `t.Fatal()` (because the Scope has the testing.T attached, we could do that). But that would not work with uncaught panics that don't go through logging. I think we really need to fix the JSON output.

In my ideal world, the Go test runner would include `"Action":"fail"` for both the unit test failures and package-level test failures. It could include the name of the package as the value for the `Test` field in that case.

Alternatively, we could find a way to ask TC to detect these package-level failures via a JSON filter.

Any case, this is not just a thing we can hand-wave away with "the tests are running concurrently, so tough". The go test runner knows there's a package that's failed and does not do its job with the `"Action"` field.

**Comment by tbg** [comment]
wait, it's not emitting a fail action for the pkg? I thought it did. It does in this simple test:

```go
func TestBoom(t *testing.T) {
	go func() {
		panic("boom")
	}()
	time.Sleep(time.Second)
}

```

```
{"Time":"2020-12-04T13:47:50.287858328+01:00","Action":"run","Package":"github.com/cockroachdb/cockroach/pkg/util/stop","Test":"TestBoom"}
{"Time":"2020-12-04T13:47:50.288002398+01:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/util/stop","Test":"TestBoom","Output":"=== RUN   TestBoom\n"}
{"Time":"2020-12-04T13:47:50.291036854+01:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/util/stop","Test":"TestBoom","Output":"panic: boom\n"}
{"Time":"2020-12-04T13:47:50.291051884+01:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/util/stop","Test":"TestBoom","Output":"\n"}
{"Time":"2020-12-04T13:47:50.291063414+01:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/util/stop","Test":"TestBoom","Output":"goroutine 31 [running]:\n"}
{"Time":"2020-12-04T13:47:50.291080754+01:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/util/stop","Test":"TestBoom","Output":"github.com/cockroachdb/cockroach/pkg/util/stop_test.TestBoom.func1()\n"}
{"Time":"2020-12-04T13:47:50.291095174+01:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/util/stop","Test":"TestBoom","Output":"\t/home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper_test.go:743 +0x39\n"}
{"Time":"2020-12-04T13:47:50.291105314+01:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/util/stop","Test":"TestBoom","Output":"created by github.com/cockroachdb/cockroach/pkg/util/stop_test.TestBoom\n"}
{"Time":"2020-12-04T13:47:50.291116554+01:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/util/stop","Test":"TestBoom","Output":"\t/home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper_test.go:742 +0x35\n"}
{"Time":"2020-12-04T13:47:50.292817397+01:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/util/stop","Output":"FAIL\tgithub.com/cockroachdb/cockroach/pkg/util/stop\t0.022s\n"}
{"Time":"2020-12-04T13:47:50.292843227+01:00","Action":"fail","Package":"github.com/cockroachdb/cockroach/pkg/util/stop","Elapsed":0.022}
```

**Comment by knz** [comment]
What is missing is a `"Test"` field

**Comment by tbg** [comment]
Ok, yes, that's what I expected.

**Comment by knz** [comment]
I'm a bit bummed that TC doesn't bother to report a package that's failed if it doesn't have also a `Test`-equipped entry inside that package that's also failed. Did it not use to do that before? I'm unsure.

**Comment by tbg** [comment]
If we're talking blame, we are blaming TC here - for not having any convenient way to look at the de-jsonified output of tests on a per-pkg basis, right? It seems that Go is "mostly doing the right thing".

**Comment by tbg** [comment]
yeah, and more broadly, that it doesn't even show the package as failed, right.

**Comment by knz** [comment]
Go could be printing a dummy `Test` entry for the last logic test that was running when the process failed.

**Comment by tbg** [comment]
No, it could not, as there are tests running concurrently. I think it is justifiable that when some random goroutine crashes, it just marks the package as failed and does not try to blame a test.

**Comment by rickystewart** [comment]
I was assigned for triage but it looks like the issue is addressed in #57557. (Feel free to re-assign/tag me if there's some other TC feature request that comes out of this.)

**Comment by knz** [comment]
The issue is being *worked around* in #57557. The workaround is unsatisfactory and janky, so we'd prefer to not go about it this way.

The core of the problem is the "Go test" module that is built in TC to parse the output of a CI command. This module has a bug (described above) and that bug needs to be fixed. There are a couple of steps from here:

1. identify where the bug is in that TC component
2. file an issue upstream
3. if the bug is not too hard to fix, propose a fix ourselves
4. track the upstream issue, and mark this issue here as closed when the upstream bug has been addressed

**Comment by rickystewart** [comment]
This is a very old bug and the logs are long-gone so it is difficult to be sure, but as far as I know, this is already fixed at `master`. Unfinished tests are generally marked with `No pass/fail/skip event for test`, and the new Bazel-based test reporting is much more granular.

**Comment by rickystewart** [comment]
I'm going to close this, although you should re-open or open a new bug if any issues still remain.
